File: engine-desktop/src/com/magma/engine/debug/modules/MapModule.java
package com.magma.engine.debug.modules;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.Dimension;
import java.util.ArrayList;

import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JComboBox;
import javax.swing.JLabel;

import com.badlogic.gdx.maps.Map;
import com.magma.engine.assets.MagmaLoader;
import com.magma.engine.maps.MapSession;
import com.magma.engine.maps.MapStage;
import com.magma.engine.stages.GameStage;
import com.magma.engine.stages.StageSwitchListener;

public class MapModule extends Module implements StageSwitchListener {

    private final JLabel current, triggerCount;
    private final JCheckBox trigger;
    private final JButton load;

    private JComboBox<String> comboBox;
    private String choosenMap; // nessesary

    private MapStage map;

    public MapModule() {
        super("Map details");
        current = new JLabel();
        triggerCount = new JLabel();

        // instantiate the widgets
        trigger = new JCheckBox("Show trigger outlines");
        trigger.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent arg0) {
                boolean on = trigger.isSelected();
                if (map != null)
                    map.setTriggersVisible(on);
            }
        });

        // load map list
        ArrayList<String> maps = MagmaLoader.getAllMapNames();
        comboBox = new JComboBox<String>();
        comboBox.setMaximumSize(new Dimension(200, 50));
        for (String map : maps) {
            comboBox.addItem(map);
        }
        add(comboBox);

        load = new JButton("Load map");
        load.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent arg0) {
                // choose the choosen map from the combobox
                choosenMap = (String) comboBox.getSelectedItem();
            }
        });

        // add the widgets to the panel
        add(current);
        add(triggerCount);
        add(trigger);
        add(load);
    }

    @Override
    public void update() {
        String curMap = "None loaded";
        current.setText(format("Loaded map", curMap));

        if (map == null)
            return;

        MapSession session = MapSession.getSession();
        if (session == null)
            return;

        int count = session.getTriggers().getCount();
        triggerCount.setText(format("Trigger count", count));

        curMap = session.getTmxName();
        current.setText(format("Loaded map", curMap));

        if (choosenMap != null) {
            // load the choosen map on the LWJGL thread
            map.requestMap(choosenMap);
            choosenMap = null;
        }
    }

    @Override
    public void stageOpened(GameStage stage) {
        if (stage instanceof MapStage) {
            map = (MapStage) stage;
        }
    }

    @Override
    public void stageClosed(GameStage stage) {
        if (stage == map) {
            map = null;
        }
    }
}

File: engine/src/com/magma/engine/debug/modules/MapModule.kt
package com.magma.engine.debug.modules

import com.magma.engine.assets.MagmaLoader.allMapNames
import com.magma.engine.maps.MapStage
import com.magma.engine.stages.GameStage
import com.magma.engine.stages.StageSwitchListener
import java.awt.Dimension
import javax.swing.JButton
import javax.swing.JCheckBox
import javax.swing.JComboBox
import javax.swing.JLabel

// TODO make for every new MapStage
class MapModule : Module("Map details"), StageSwitchListener {
    private val current: JLabel = JLabel()
    private val triggerCount: JLabel = JLabel()
    private val trigger: JCheckBox = JCheckBox("Show trigger outlines")
    private val load: JButton = JButton("Load map")
    private val comboBox: JComboBox<String>

    private var chosenMap : String? = null // must be deferred
    private var mapStage: MapStage? = null

    init {
        // instantiate the widgets
        trigger.addActionListener {
            val on = trigger.isSelected
            mapStage?.setTriggersVisible(on)
        }

        // load map list
        val maps = allMapNames
        comboBox = JComboBox()
        comboBox.maximumSize = Dimension(200, 50)
        for (map in maps) {
            comboBox.addItem(map)
        }
        add(comboBox)

        load.addActionListener { // choose the choosen map from the combobox
            chosenMap = comboBox.selectedItem as String
        }

        // add the widgets to the panel
        add(current)
        add(triggerCount)
        add(trigger)
        add(load)
    }

    override fun update() {
        var curMap = "None loaded"
        current.text = format("Loaded map", curMap)
        if (mapStage == null) return
        val count: Int = mapStage!!.triggers.count
        triggerCount.text = format("Trigger count", count)
        curMap = mapStage!!.tmxName
        current.text = format("Loaded map", curMap)
        if (chosenMap != null) {
            // load the choosen map on the LWJGL thread
            mapStage!!.unloadMap()
            mapStage!!.loadMap(chosenMap!!)
            chosenMap = null
        }
    }

    override fun stageOpened(stage: GameStage) {
        if (stage is MapStage) {
            mapStage = stage
        }
    }

    override fun stageClosed(stage: GameStage) {
        if (stage === mapStage) {
            mapStage = null
        }
    }
}